### PR TITLE
pre-commit hook to avoid common mistakes

### PR DESCRIPTION
**Pull request by makarandp0**
Yesterday I commited a change which marked a test suite  `.only` - I ended up cutting a rc with it.
This change reverts  the `.only` and adds a pre-commit hook to avoid such errors in future. The pre-commit hook is not on by default - instructions to enable it are added to readme.

**Contributing to Twilio**

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

File: pre-commit.sh [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    echo "COMMIT REJECTED: found '$FORBIDDEN_PRINTABLE' references. Please remove them before committing."
+    exit 1
+  fi
+}
+
+echo "Validating current branch..."
+validate_branch
+
+echo "Checking for .only..."
+check_diffs 'test/unit/spec/' '\.js$' '\.only' '.only'
```
Comment (makarandp0) [inline pre-commit.sh:55]: should include integration tests as well.

**Comment by manjeshbhargav** [comment]
@makarandp0 Let's hold off on doing this for now. Please create a ticket for adding a pre-commit hook for running lint and unit tests + checking `.only` is not specified.

**Comment by makarandp0** [comment]
@manjeshbhargav - I can file ticket to add a more comprehensive pre-commit hooks, but for I think whatever hooks we have here can help immediately, and should be merged.

**Comment by makarandp0** [comment]
btw the pre-hooks here are optional, This change does not add a pre-commit hook for the repo, its just adds a hook, that developers can optionally add by following instructions in readme.

File: pre-commit.sh [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    echo "COMMIT REJECTED: found '$FORBIDDEN_PRINTABLE' references. Please remove them before committing."
+    exit 1
+  fi
+}
+
+echo "Validating current branch..."
+validate_branch
+
+echo "Checking for .only..."
+check_diffs 'test/' '\.js$' '\.only' '.only'
```
Comment (manjeshbhargav) [inline pre-commit.sh:55]:
```suggestion
check_diffs 'test/' '\.js$' '\.only(' '.only('
```

File: pre-commit.sh [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+# it wants to stop the commit.
+#
+
+# Check 1: check for direct commit against protected branches.
+# ------------------------------------------------------------
+function validate_branch () {
+    # notice current_branch evaluation only looks at part at the end of
+    # so if you want to exclude feature/foo, list foo in the protected_branches
+
+    PROTECTED_BRANCHES=('master' 'develop' 'support-1.x')
```
Comment (manjeshbhargav) [inline pre-commit.sh:15]: `develop` should not be a protected branch.
```suggestion
    PROTECTED_BRANCHES=('master' 'support-1.x')
```

File: pre-commit.sh [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+#
+# hook script to verify what is about to be committed.
+# Called by "git commit" with no arguments.  The hook should
+# exit with non-zero status after issuing an appropriate message if
+# it wants to stop the commit.
+#
+
+# Check 1: check for direct commit against protected branches.
+# ------------------------------------------------------------
+function validate_branch () {
```
Comment (manjeshbhargav) [inline pre-commit.sh:11]:
```suggestion
function block_protected_branch_commits () {
```

File: pre-commit.sh [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+            echo "You were about to commit to a forbidden branch: $i"
+            echo "Commit aborted"
+            exit 1;
+        fi
+    done
+}
+
+# Check 2: Ensure that you are not commiting any tests with marked .only
+# ----------------------------------------------------------------------
+function check_diffs () {
```
Comment (manjeshbhargav) [inline pre-commit.sh:30]:
```suggestion
function block_commit_for_diff_patterns () {
```

File: pre-commit.sh (hunk 3 above)
Comment (makarandp0) [reply pre-commit.sh:15]: why isn't develop protected ? in what cases do we commit to it directly? if it's only very rare occurrences I recommend using `git commit --no-verify` than opening up  `develop` here.

**Review by manjeshbhargav — APPROVED** [review]

**Pull request merged**